src/solver/clause.rs

```diff
@@ -356,249 +379,204 @@ impl ClauseState {
-                .cloned()
-                .find(|&l| can_watch(l))
-                .map(|l| l.solvable_id),
-            Clause::Constrains(..) | Clause::ForbidMultipleInstances(..) | Clause::Lock(..) => None,
-            Clause::Requires(solvable_id, requirement) => {
-                // The solvable that added this clause
-                let solvable_lit = Literal {
-                    solvable_id,
-                    negate: true,
-                };
-                if can_watch(solvable_lit) {
-                    return Some(solvable_id);
-                }
-
-                // The available candidates
-                for &candidate in &requirement_to_sorted_candidates[&requirement] {
-                    let lit = Literal {
-                        solvable_id: candidate.into(),
-                        negate: false,
-                    };
-                    if can_watch(lit) {
-                        return Some(candidate.into());
-                    }
-                }
-
-                // No solvable available to watch
+            Clause::Constrains(..) | Clause::ForbidMultipleInstances(..) | Clause::Lock(..) => {
+                // We cannot move the watches in these clauses.
                 None
             }
+            clause => {
+                let next = clause.try_fold_literals(
+                    learnt_clauses,
+                    requirement_to_sorted_candidates,
+                    (),
+                    |_, lit| {
+                        // The next unwatched variable (if available), is a variable that is:
+                        // * Not already being watched
+                        // * Not yet decided, or decided in such a way that the literal yields true
+                        if self.watched_literals[other_watch_index].solvable_id()
+                            != lit.solvable_id()
+                            && lit.eval(decision_map).unwrap_or(true)
+                        {
+                            ControlFlow::Break(lit)
+                        } else {
+                            ControlFlow::Continue(())
+                        }
+                    },
+                );
+                match next {
+                    ControlFlow::Break(lit) => Some(lit),
+                    ControlFlow::Continue(_) => None,
+                }
+            }
         }
     }
 }
 
 /// Represents a literal in a SAT clause (i.e. either A or ¬A)
-#[derive(Copy, Clone, Eq, PartialEq, Hash)]
-pub(crate) struct Literal {
-    pub(crate) solvable_id: InternalSolvableId,
-    pub(crate) negate: bool,
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub(crate) struct Literal(u32);
+
+impl Literal {
+    /// Constructs a new [`Literal`] from a [`InternalSolvableId`] and a boolean
+    /// indicating whether the literal should be negated.
+    pub fn new(solvable_id: InternalSolvableId, negate: bool) -> Self {
+        assert!(solvable_id.0 < (u32::MAX >> 1) - 1, "solvable id too big");
```

Comment:
Could / should this be a debug assert? I assume we're not constructing so many literals on the fly anymore, so perhaps this is fine?